src/Chainweb/Pact/PactService.hs

```diff
@@ -1074,7 +1144,7 @@ execLocal
     :: PayloadCasLookup cas
     => ChainwebTransaction
     -> PactServiceM cas (P.CommandResult P.Hash)
-execLocal cmd = do
+execLocal cmd = withDiscardedBatch $ do
```

Comment:
`_cpDiscard` discards only `Block` SAVEPOINTs. It doesn't discard the `PreBlock` SAVEPOINTS.

`_cpRestore` does the following SAVEPOINT operations:

```
1.  open `PreBlock`
2. release/rollback `PreBlock`
3. open `Block`
```

`_cpDiscard` does

```
4. release/rollback `Block`
```

The `Batch` SAVEPOINTs wrap all of the above. Without it aborting within the action given to `withCheckpointer` would only rollback `Block` but not `PreBlock`.

Reply:
In particular `_cpSave` is a complex operation that even when executed under a mask in a finalizer may fail in different ways. If that happens, it would result in a leaked `SAVEPOINT`. Discarding a batch is much simpler and more robust.

Reply:
I think the `PreBlock` was originally intended to possibly "keep the rewind but discard the playthrough" on new block? In any case, I think we should ditch the `PreBlock` batch altogether, since it should succeed or fail alongside `Block`. Not for this PR obviously.

Reply:
Agreed, that this is stuff for another PR. This PR tries to be conservative with respect to lower-level checkpointer and pact execution logic.